core/trino-main/src/main/java/io/trino/operator/ScanFilterAndProjectOperator.java

```diff
@@ -440,13 +441,13 @@ public ScanFilterAndProjectOperatorFactory(
             this.cursorProcessor = requireNonNull(cursorProcessor, "cursorProcessor is null");
             this.pageProcessor = requireNonNull(pageProcessor, "pageProcessor is null");
             this.sourceId = requireNonNull(sourceId, "sourceId is null");
-            this.pageSourceProvider = requireNonNull(pageSourceProvider, "pageSourceProvider is null");
             this.table = requireNonNull(table, "table is null");
             this.columns = ImmutableList.copyOf(requireNonNull(columns, "columns is null"));
             this.dynamicFilter = dynamicFilter;
             this.types = requireNonNull(types, "types is null");
             this.minOutputPageSize = requireNonNull(minOutputPageSize, "minOutputPageSize is null");
             this.minOutputPageRowCount = minOutputPageRowCount;
+            this.pageSourceProvider = pageSourceProvider.createPageSourceProvider(table.catalogHandle());
```

Comment:
nit: these can be created in `LocalExecutionPlanner` instead. There are queries which would run multiple table scans in same source stage (e.g. union all queries). It would be beneficial if scanned tables are the same. However it might retain more memory though if tables are different.

Reply:
Maybe.  This Iceberg code currently does not support having multiple different table scans in the same provider, so that would need to be changed first.  Also, Iceberg doesn't get any benefit from cross table scan (except the more rare self colocated join on a table case)